rfd/0001-runjob-design.md

```diff
@@ -0,0 +1,288 @@
+      --cpu-limit=1000                        Limit of CPU usage allowed in millis. 1000 = 1 CPU
+      --memory-limit=268435456                Limit of memory allowed in bytes
+      --process-limit=5                       Maximum number of child processes allowed
+      --enable-network                        Enable network access for the process
+      --iops-limit=900000000                  Maximum number of read/write IO per second
+  -A, --command-args=COMMAND-ARGS,...         Arguments to the command being ran
+```
+
+Example of starting a job:
+
+```
+$ runjob start "/bin/bash" \
+-A "echo hello world" \
+--server-host="localhost:8443" \
+--cert-ca-path="certs/cert-ca.crt" \
+--tls-cert-path="certs/user-tls.crt" \
+--tls-key-path="certs/user-tls.key" \
+--server-command="/bin/bash" \
+```
+
+
+### Job Resource Limits
+
+Resource limits must be specified when creating a job. To be explicit, these are limits rather than allocations.
+
+These limits will be implemented in the library, using the Linux cgroup v2 functionality, which will use its nomenclature for specifying limits.
+
+- CPU Time - in Millis; 1000 millis = 1 CPU
+- Memory Limit - in bytes
+- Max Processes - the number of processes a parent process can fork
+- IOPS - the number of read and write operations per second
+
+In the future, IOPS should be more granular to support separate configurations for read, write, and the read/write BPS instead of just the number of operations.
+
+There will not be any limits on the number of jobs a user can run, how many resources they use, etc.  Nor will there be any support for process monitoring or restarting.  At least some basic level of not allowing more than `n` processes per user seems like an important feature to implement in the next release.
+
+### Job Isolation
+
+The library will create the isolation before executing the Job.
+
+Each Job will run in its own PID and mount namespace.  This ensures that it only has access to itself and its children and that other processes also can't kill the processes in this namespace. Requested Jobs cannot see other PIDs or mounts running on the system.
+
+By default, each Job will also have its own network namespace to prevent sending or receiving traffic across the Internet or local network.  This ensures a user can only run a random process and not import from nor export any data to other network and internet resources.
+
+A user will be able to specify that networking is enabled for a job. If enabled, the Job will then be able to use the same network interface as its parent process.
+
+The functionality of network namespaces should be implemented further in a later version of the product to allow more flexibility, such as allowing a user to define which network interfaces should be in the container.
+
+### Job Execution
+
+The library will use a separate executable, `jobrunner,` to run the jobs. This runner will circumvent issues with using forking in Go.
```

Comment:
Makes sense and thanks for pointing it out. I was conflating it with being about 3rd party binaries rather than ones built as part of the project itself.

After thinking about it for awhile, I think I understand how to make this work as far as getting around forking is concerned.  Executing the same binary recursively.  Painfully obvious now. :smiley:

Essentially make the server binary have different commands; one is to serve requests, the other to create the PID, namespace, process group, etc, then start the command from within it. So basically the server execs its own binary again but in a different mode.  

E.g:

`$ jobmanager serve --listen localhost:8080`

Which then uses itself to create isolation/resource limits and then starts the command, pipes output back to the server process:

`$ jobmanger exec --command /bin/foo --args foo `

Is that more on the right path?  

I understand that I probably also need more detail around using syscall, process groups, etc.  Trying to figure out how to get around forking without another binary was where I was stuck.

Reply:
The Job Execution execution is now updated and much more thorough.